File: backends/spring-java/src/main/java/dev/stackverse/backend/web/MetaController.java
package dev.stackverse.backend.web;

import java.util.Map;
import org.springframework.dao.DataAccessException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.simple.JdbcClient;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class MetaController {
    private final JdbcClient jdbcClient;

    public MetaController(JdbcClient jdbcClient) {
        this.jdbcClient = jdbcClient;
    }

    @GetMapping("/healthz")
    public Map<String, String> healthz() {
        return Map.of("status", "up");
    }

    @GetMapping("/readyz")
    public ResponseEntity<Map<String, String>> readyz() {
        try {
            jdbcClient.sql("select 1").query(Integer.class).single();
            return ResponseEntity.ok(Map.of("status", "ready"));
        } catch (DataAccessException exception) {
            return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE).body(Map.of("status", "unavailable"));
        }
    }
}

File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/web/MetaController.kt
package dev.stackverse.backend.web

import org.springframework.dao.DataAccessException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.jdbc.core.simple.JdbcClient
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController

/** Liveness/readiness for the container runtime; not proxied by the gateway. */
@RestController
class MetaController(private val jdbcClient: JdbcClient) {

    @GetMapping("/healthz")
    fun healthz(): Map<String, String> = mapOf("status" to "up")

    @GetMapping("/readyz")
    fun readyz(): ResponseEntity<Map<String, String>> = try {
        jdbcClient.sql("select 1").query(Int::class.java).single()
        ResponseEntity.ok(mapOf("status" to "ready"))
    } catch (e: DataAccessException) {
        ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE).body(mapOf("status" to "unavailable"))
    }
}
